router/batchrouter/asyncdestinationmanager/sftp/utils.go

```diff
@@ -0,0 +1,313 @@
+package sftp
+
+import (
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	jsoniter "github.com/json-iterator/go"
+	sftp "github.com/rudderlabs/rudder-go-kit/sftp"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+// createSSHConfig creates SSH configuration based on destination
+func createSSHConfig(destination *backendconfig.DestinationT) (*sftp.SSHConfig, error) {
+	destinationConfigJson, err := json.Marshal(destination.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	var config destConfig
+	if err := json.Unmarshal(destinationConfigJson, &config); err != nil {
+		return nil, err
+	}
+
+	if err := validate(config); err != nil {
+		return nil, err
+	}
+	port, _ := strconv.Atoi(config.Port)
+	sshConfig := &sftp.SSHConfig{
+		User:       config.Username,
+		HostName:   config.Host,
+		Port:       port,
+		AuthMethod: config.AuthMethod,
+		Password:   config.Password,
+		PrivateKey: config.PrivateKey,
+	}
+
+	return sshConfig, nil
+}
+
+// getFieldNames extracts the field names from the first JSON record.
+func getFieldNames(records []record) ([]string, error) {
+	if len(records) == 0 {
+		return nil, fmt.Errorf("no records found")
+	}
+	fields, ok := records[0]["message"].(map[string]interface{})["fields"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("fields not found in the first record")
```

Comment:
Once this error propagates and ends up in the logs, do we have a way to tell which record was this? Maybe we can add more info to the error?

Reply:
This is very corner case. Sources always send `fields` as they are required for setting up the sync. Perhaps we can consider ignoring this error